org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ADTreeOnDropListener.java

```diff
@@ -170,15 +175,20 @@ private void moveNode(DefaultTreeNode<Object> movingNode, DefaultTreeNode<Object
 			{
 				DefaultTreeNode<?> nd = (DefaultTreeNode<?>)oldParent.getChildAt(i);
 				MTreeNode md = (MTreeNode) nd.getData();
-				StringBuilder sql = new StringBuilder("UPDATE ");
-				sql.append(mTree.getNodeTableName())
-					.append(" SET Parent_ID=").append(oldMParent.getNode_ID())
-					.append(", SeqNo=").append(i)
-					.append(", Updated=getDate()")
-					.append(" WHERE AD_Tree_ID=").append(mTree.getAD_Tree_ID())
-					.append(" AND Node_ID=").append(md.getNode_ID());
-				if (log.isLoggable(Level.FINE)) log.fine(sql.toString());
-				no = DB.executeUpdate(sql.toString(),trx.getTrxName());
+				if (md.getNode_ID() == mtnMovingNode.getNode_ID()) {
+					updateMovingNodePO(oldMParent, mtnMovingNode, i, trx.getTrxName());
+				}
+				else {
+					StringBuilder sql = new StringBuilder("UPDATE ");
+					sql.append(mTree.getNodeTableName())
+						.append(" SET Parent_ID=").append(oldMParent.getNode_ID())
+						.append(", SeqNo=").append(i)
+						.append(", Updated=getDate()")
+						.append(" WHERE AD_Tree_ID=").append(mTree.getAD_Tree_ID())
+						.append(" AND Node_ID=").append(md.getNode_ID());
+					if (log.isLoggable(Level.FINE)) log.fine(sql.toString());
+					no = DB.executeUpdate(sql.toString(),trx.getTrxName());
+				}

```

Comment:
this is both incomplete and confusing. incomplete as you still have a mix of po and sql update. confusing as why the need to have mtnMovingNode and the added condition? shouldn't it be just a direct replacement of sql script update with PO update without the added variables and condition ?

Reply:
> this is both incomplete and confusing. incomplete as you still have a mix of po and sql update. confusing as why the need to have mtnMovingNode and the added condition? shouldn't it be just a direct replacement of sql script update with PO update without the added variables and condition ?

The update is performed in a loop - that is, for all nodes of the parent to which the movingNode belongs. If I replace SQL update with PO then change events will be generated for all nodes - but my goal is to generate the change event only for the node being moved, therefore I added a condition to check that the current node that will be updated is equal to the moving node and only in this case I use the PO.
Please correct me if I'm going in the wrong direction.
![updatePO](https://user-images.githubusercontent.com/87425658/128363203-df5c4fcc-a1de-462b-890c-90b083e870e2.png)

Reply:
I think it is fine to use PO update for all node that's being updated here. It is perhaps slower but it is more consistent in that we have event and change log for all node that have been updated here.

Reply:
Agree, but how we can provide event generation ONLY for movingNode ?

Reply:
You don't have to. In the event handler, you can determine whether the PO is the movingNode by checking which value of the PO have been changed.